### ISSUE TITLE
Phase III Lab/Experiment(s) OnBoarding Request for <fill your lab name here>

### ISSUE THREAD

**Issue by amritamindbrain** [virtual-labs/engineers-forum#682]
#### 
**Lab Name**:  Bioinformatics and Data Science in Biotechnology Lab
**Department**: Biotechnology and Biomedical Engineering

<br>

**List of Experiments and Repositories**: 

1: Writing and reading sequence data in R  -  https://github.com/virtual-labs/exp-sequence-data-r-au,  Master Branch, v1.0

2: Guanine-Cytosine content analysis and basics of DNA sequence statistics   -  https://github.com/virtual-labs/exp-dna-sequence-statistics-au,  Master Branch, v1.0

3: Pairwise sequence alignment of Protein or DNA sequences  -  https://github.com/virtual-labs/exp-pairwise-sequence-alignment-au,   Master Branch, v1.0

4: DNA sequence analysis and determining DNA open reading frames  -  https://github.com/virtual-labs/exp-open-reading-frames-au,   Master Branch, v1.0

<br>

**List of proposed experiments to be updated under Bioinformatics and Data Science in Biotechnology**: *(require server support)*
1: Reading Fasta using SequinR - https://github.com/virtual-labs/exp-sequinr-au
2: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au
3: UniProt Protein Sequence Retrieval in R - https://github.com/virtual-labs/exp-sequence-retrieval-au
4: Retrieving a list of sequences from UniProt - https://github.com/virtual-labs/exp-uniprot-au
5: Computing scoring matrices for amino acids and long pairwise alignment in R - https://github.com/virtual-labs/exp-long-pairwise-alignment-au
6: Calculating genetic distances between protein sequences - https://github.com/virtual-labs/exp-protein-sequences-au
7: Gene finding: Finding Start and Stop codons using R - https://github.com/virtual-labs/exp-gene-finding-au
8: Differential expression analyses of RNA-seq - https://github.com/virtual-labs/exp-rna-seq-au

#### 

**Introduction:**

In this lab, students will be able to learn and practice basics of R programming. Students can expand their skillset in R programming by reading and writing biological sequence data in R platform. Educating this will allow users to learn how to use R as an open source language for learning bioinformatics data processing. Specifically, this lab will help analyse biological sequence data using simple R code snippets. With the features such as flexibility, data handling and modelling capabilitie R programming has been a widely accepted software tools for bioinformatics. It also supports the creation and use of self-describing data structures. With the trends in data science, the lab focuses on FOSS implementations which can be further utilized as laboratory manual or reference material for life science students. 

**Objectives**
1.	Introduction to the basics of R programming techniques in bioinformatics and bioscience data processing. 
2.	 Retrieving public sequence databases
3.	Use of computational tools to find sequences, analysis of protein and amino acid sequences without the need of many other software packages. 
4.	To retrieve biological sequence of different organisms and to identify regions for similarity for phylogenic analysis.
5.	To study sequence alignment methods using R code snippets even for students without biology background.

**Course Alignment:**
UG in Biotechnology/Microbiology/Bioinformatics
PG in Biotechnology/Microbiology/Bioinformatics
B.Tech./ B.E in Biotechnology

**Universities Mapped :**
Anna University, Tamil Nadu
Visvesvaraya Technological University, Karnataka
Bharathiyar University, Tamil Nadu
Bharathidasan University, Tamil Nadu
Madurai Kamaraj University, Tamil Nadu
Savitribai Phule Pune University, Maharashtra

**Target Audience:**
(1) UG
(2) PG
(3) B.Tech./ B.E
(4) PhD/Research Scholars

**Comment by pavanchow** [comment]
@amritamindbrain 
1. Considering this is an R based lab, could you please share your hardware requirements for the server? (ie., RAM, no of processors, etc..,) 
2. We are hosting all the other labs with an apache server, are these R based applications complaint with apache, or need any other specific web server?

**Comment by pavanchow** [comment]
Could you please follow .MD format while updating the .md files in [experiment](https://github.com/virtual-labs/exp-sequence-data-r-au/tree/main/experiment) directory of the experiments. The title of the experiment is broken in [testing links](https://virtual-labs.github.io/exp-sequence-data-r-au/) as ## are removed from the md file. 


![image](https://user-images.githubusercontent.com/24794698/109453506-422f6a80-7a78-11eb-9de5-98305f5996ce.png)

**Comment by pavanchow** [comment]
@amritamindbrain @joshy-alphonse
 Could please use semantic versioning for tagging? can you please create **tag v1.0.0** for requested experiments,

**Comment by joshy-alphonse** [comment]
> @amritamindbrain
> 
> 1. Considering this is an R based lab, could you please share your hardware requirements for the server? (ie., RAM, no of processors, etc..,)
> 2. We are hosting all the other labs with an apache server, are these R based applications complaint with apache, or need any other specific web server?


@pavanchow  The [DockerFile](https://github.com/virtual-labs/engineers-forum/files/6066138/Rdocker.zip)  for the R based labs.

one more thing i would like to know is there some kind of load balancer in your end or we need to provide our own in the container itself?

**Comment by raj-vlabs** [comment]
@joshy-alphonse 
With the setup that we discussed, I am not sure what you can achieve with an in the container load balancer. Since the container runs as a single process, there is no point in doing any load balancing inside it. If you want load balancing to be done, it has to be done outside of the container.
In my view, you do not need load balancing at the moment. The traffic on the lab will slowly increase with time. At that point we may need to either commission a more powerful machine or multiple machines with a load balancer. In the latter case, you have to ensure that the user gets a consistent experience even when one of the machines go down.
In short, I think the application does not need a load balancer right now and some work will be required if we want to make it load balanced.
Please correct me if I didn't understand something correctly.

**Comment by joshy-alphonse** [comment]
@raj-vlabs  yes a load balancer within the container wont make sense in that scenario...

**Comment by amritamindbrain** [comment]
**Lab Name:** Bioinformatics and Data Science in Biotechnology Lab
**Department**: Biotechnology and Biomedical Engineering


**List of Experiments and Repositories:**

1: Writing and reading sequence data in R - https://github.com/virtual-labs/exp-sequence-data-r-au, Master Branch, v1.0.0

2: Guanine-Cytosine content analysis and basics of DNA sequence statistics - https://github.com/virtual-labs/exp-dna-sequence-statistics-au, Master Branch, v1.0.0

3: Pairwise sequence alignment of Protein or DNA sequences - https://github.com/virtual-labs/exp-pairwise-sequence-alignment-au, Master Branch, v1.0.0

4: DNA sequence analysis and determining DNA open reading frames - https://github.com/virtual-labs/exp-open-reading-frames-au, Master Branch, v1.0.0

5: Reading Fasta using SequinR - https://github.com/virtual-labs/exp-sequinr-au, Master Branch, v1.0.0

6: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au, Master Branch, v1.0.0

7: UniProt Protein Sequence Retrieval in R - https://github.com/virtual-labs/exp-sequence-retrieval-au, Master Branch, v1.0.0

8:  Gene finding: Finding Start and Stop codons using R - https://github.com/virtual-labs/exp-gene-finding-au, Master Branch, v1.0.0


<br>

**List of proposed experiments to be updated under Bioinformatics and Data Science in Biotechnology:** _(require server support)_
1: Retrieving a list of sequences from UniProt - https://github.com/virtual-labs/exp-uniprot-au
2: Computing scoring matrices for amino acids and long pairwise alignment in R - https://github.com/virtual-labs/exp-long-pairwise-alignment-au
3: Calculating genetic distances between protein sequences - https://github.com/virtual-labs/exp-protein-sequences-au
4: Differential expression analyses of RNA-seq - https://github.com/virtual-labs/exp-rna-seq-au

<br>

**Introduction:**
In this lab, students will be able to learn and practice basics of R programming. Students can expand their skillset in R programming by reading and writing biological sequence data in R platform. Educating this will allow users to learn how to use R as an open source language for learning bioinformatics data processing. Specifically, this lab will help analyse biological sequence data using simple R code snippets. With the features such as flexibility, data handling and modelling capabilitie R programming has been a widely accepted software tools for bioinformatics. It also supports the creation and use of self-describing data structures. With the trends in data science, the lab focuses on FOSS implementations which can be further utilized as laboratory manual or reference material for life science students.

<br>

**Objectives**
Introduction to the basics of R programming techniques in bioinformatics and bioscience data processing.
Retrieving public sequence databases
Use of computational tools to find sequences, analysis of protein and amino acid sequences without the need of many other software packages.
To retrieve biological sequence of different organisms and to identify regions for similarity for phylogenic analysis.
To study sequence alignment methods using R code snippets even for students without biology background.

<br>

**Course Alignment:**
UG in Biotechnology/Microbiology/Bioinformatics
PG in Biotechnology/Microbiology/Bioinformatics
B.Tech./ B.E in Biotechnology

<br>

**Universities Mapped :**
Anna University, Tamil Nadu
Visvesvaraya Technological University, Karnataka
Bharathiyar University, Tamil Nadu
Bharathidasan University, Tamil Nadu
Madurai Kamaraj University, Tamil Nadu
Savitribai Phule Pune University, Maharashtra

<br>
**Target Audience:**

(1) UG
(2) PG
(3) B.Tech./ B.E
(4) PhD/Research Scholars

**Comment by pavanchow** [comment]
@amritamindbrain 

Hosted Bioinformatics and Data Science in Biotechnology Lab with JavaScript experiments 
URL of Lab: https://bds-au.vlabs.ac.in/
Branch: Master

Please check and approve

**Comment by amritamindbrain** [comment]
**Lab Name:** Bioinformatics and Data Science in Biotechnology Lab
**Department**: Biotechnology and Biomedical Engineering


**List of Experiments and Repositories:**

1: Writing and reading sequence data in R - https://github.com/virtual-labs/exp-sequence-data-r-au, Master Branch, v1.0.1

2: Guanine-Cytosine content analysis and basics of DNA sequence statistics - https://github.com/virtual-labs/exp-dna-sequence-statistics-au, Master Branch, v1.0.0

3: Pairwise sequence alignment of Protein or DNA sequences - https://github.com/virtual-labs/exp-pairwise-sequence-alignment-au, Master Branch, v1.0.0

4: DNA sequence analysis and determining DNA open reading frames - https://github.com/virtual-labs/exp-open-reading-frames-au, Master Branch, v1.0.0

5: Reading Fasta using SequinR - https://github.com/virtual-labs/exp-sequinr-au, Master Branch, v1.0.0

6: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au, Master Branch, v1.0.0

7: UniProt Protein Sequence Retrieval in R - https://github.com/virtual-labs/exp-sequence-retrieval-au, Master Branch, v1.0.0

8:  Gene finding: Finding Start and Stop codons using R - https://github.com/virtual-labs/exp-gene-finding-au, Master Branch, v1.0.0


<br>

**List of proposed experiments to be updated under Bioinformatics and Data Science in Biotechnology:** _(require server support)_
1: Retrieving a list of sequences from UniProt - https://github.com/virtual-labs/exp-uniprot-au
2: Computing scoring matrices for amino acids and long pairwise alignment in R - https://github.com/virtual-labs/exp-long-pairwise-alignment-au
3: Calculating genetic distances between protein sequences - https://github.com/virtual-labs/exp-protein-sequences-au
4: Differential expression analyses of RNA-seq - https://github.com/virtual-labs/exp-rna-seq-au

<br>

**Introduction:**
In this lab, students will be able to learn and practice basics of R programming. Students can expand their skillset in R programming by reading and writing biological sequence data in R platform. Educating this will allow users to learn how to use R as an open source language for learning bioinformatics data processing. Specifically, this lab will help analyse biological sequence data using simple R code snippets. With the features such as flexibility, data handling and modelling capabilitie R programming has been a widely accepted software tools for bioinformatics. It also supports the creation and use of self-describing data structures. With the trends in data science, the lab focuses on FOSS implementations which can be further utilized as laboratory manual or reference material for life science students.

<br>

**Objectives**

- Introduction to the basics of R programming techniques in bioinformatics and bioscience data processing.
- Retrieving public sequence databases
- Use of computational tools to find sequences, analysis of protein and amino acid sequences without the need of many other software packages.
- To retrieve biological sequence of different organisms and to identify regions for similarity for phylogenic analysis.
- To study sequence alignment methods using R code snippets even for students without biology background.

<br>

**Course Alignment:**

- UG in Biotechnology/Microbiology/Bioinformatics
- PG in Biotechnology/Microbiology/Bioinformatics
- B.Tech./ B.E in Biotechnology

<br>

**Universities Mapped :**

- Anna University, Tamil Nadu
- Visvesvaraya Technological University, Karnataka
- Bharathiyar University, Tamil Nadu
- Bharathidasan University, Tamil Nadu
- Madurai Kamaraj University, Tamil Nadu
- Savitribai Phule Pune University, Maharashtra

<br>
**Target Audience:**

(1) UG
(2) PG
(3) B.Tech./ B.E
(4) PhD/Research Scholars

**Comment by amritamindbrain** [comment]
Some correction have been done and are now ready to be hosted

**Comment by pavanchow** [comment]
@amritamindbrain You don't need to mention Introduction, Objectives, Course Alignment, Universities, and Target Audience on every rehosting request.

**Comment by amritamindbrain** [comment]
Some changes have been made in the objectives too.

**Comment by amritamindbrain** [comment]
> @amritamindbrain You don't need to mention Introduction, Objectives, Course Alignment, Universities, and Target Audience on every rehosting request.

Some changes (added bullets) have been made in Objectives, Course Alignment, Universities, and Target Audience

**Comment by pavanchow** [comment]
@amritamindbrain

Hosted Bioinformatics and Data Science in Biotechnology Lab with JavaScript experiments
URL of Lab: https://bds-au.vlabs.ac.in/
Branch: Master

Please check and approve

**Comment by amritamindbrain** [comment]
For 2 experiments, some of the corrections are not reflected in the given URL.

1: Writing and reading sequence data in R - https://github.com/virtual-labs/exp-sequence-data-r-au, Master Branch, v1.0.1
6: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au, Master Branch, v1.0.0

**Comment by pavanchow** [comment]
@amritamindbrain 
The below experiment has commits on dev and testing branches but not pushed to the main branch 
6: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au

**Comment by amritamindbrain** [comment]
> @amritamindbrain
> The below experiment has commits on dev and testing branches but not pushed to the main branch
> 6: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au

updated

**Comment by pavanchow** [comment]
@amritamindbrain 
Could you please create a tag for "6: Querying NCBI database in R" as experiments are deployed only using tags. Thanks.

**Comment by amritamindbrain** [comment]
> @amritamindbrain
> Could you please create a tag for "6: Querying NCBI database in R" as experiments are deployed only using tags. Thanks.

6: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au, Master Branch, v1.0.0

**Comment by pavanchow** [comment]
@amritamindbrain

Hosted Bioinformatics and Data Science in Biotechnology Lab with above experiments
URL of Lab: https://bds-au.vlabs.ac.in/
Branch: Master

Please check

**Comment by amritamindbrain** [comment]
Everything fine except
1: Writing and reading sequence data in R - https://github.com/virtual-labs/exp-sequence-data-r-au, Master Branch, v1.0.1
 The corrections made are not reflected in the given URL.

**Comment by pavanchow** [comment]
@amritamindbrain The testing branch is 3 commits ahead of the main branch. Can you please merge testing into the main and create v1.0.2 tag. 
![image](https://user-images.githubusercontent.com/24794698/110689892-40e50700-8209-11eb-9052-1265718206c5.png)

**Comment by joshy-alphonse** [comment]
@pavanchow  the branches have been merged and a new updated release tag with v1.0.2 is published.

**Comment by pavanchow** [comment]
@amritamindbrain

Hosted Bioinformatics and Data Science in Biotechnology Lab with 1: Writing and reading sequence data in R 
URL of Lab: https://bds-au.vlabs.ac.in/
Branch: Master

Please check and approve

**Comment by amritamindbrain** [comment]
verified. Lets go ahead.

**Comment by raj-vlabs** [comment]
> > @amritamindbrain
> > 
> > 1. Considering this is an R based lab, could you please share your hardware requirements for the server? (ie., RAM, no of processors, etc..,)
> > 2. We are hosting all the other labs with an apache server, are these R based applications complaint with apache, or need any other specific web server?
> 
> @pavanchow The [DockerFile](https://github.com/virtual-labs/engineers-forum/files/6066138/Rdocker.zip) for the R based labs.
> 
> one more thing i would like to know is there some kind of load balancer in your end or we need to provide our own in the container itself?

@amritamindbrain @joshy-alphonse 
I read the Dockerfile. I have a couple of questions and suggestions:

1. The Dockerfile cannot be shared by attaching to an issue. That will make version control impossible. This needs to go into a repository and properly tagged before it can be deployed.
2. Since the Dockerfile is used to build a container which will host all the labs, this cannot live in the repo of an experiment or lab. We should probably create a separate repo for this. We need to discuss this before coming to a conclusion.
3. The Dockerfile does not load the code for the experiments anywhere. In your local setup this may be working because docker copies the current directory to the container by default. In the production environment, since this file will be in its own repo and this repo will not have any of the experiment code, this will not work. Please correct me if I am wrong but I think you need to explicitly load the labs to be deployed to the container. That will also help version control and better management of deployments.
Please feel free to discuss the points which are incorrect or unclear.

best regards
raj

**Comment by BalammaBoya** [comment]
**Lab Name:** Bioinformatics and Data Science in Biotechnology Lab
**Department**: Biotechnology and Biomedical Engineering

This issue also is used to host labs and their experiments that have been migrated to Pretest and Posttest questions ".js" format to ".json" format and updated .yml file.


**List of Experiments and Repositories:**

1: Writing and reading sequence data in R - https://github.com/virtual-labs/exp-sequence-data-r-au, Master Branch, v1.0.3

2: Guanine-Cytosine content analysis and basics of DNA sequence statistics - https://github.com/virtual-labs/exp-dna-sequence-statistics-au, Master Branch, v1.0.1

3: Pairwise sequence alignment of Protein or DNA sequences - https://github.com/virtual-labs/exp-pairwise-sequence-alignment-au, Master Branch, v1.0.1

4: DNA sequence analysis and determining DNA open reading frames - https://github.com/virtual-labs/exp-open-reading-frames-au, Master Branch, v1.0.1

5: Reading Fasta using SequinR - https://github.com/virtual-labs/exp-sequinr-au, Master Branch, v1.0.1

6: Querying NCBI database in R - https://github.com/virtual-labs/exp-querying-ncbi-au, Master Branch, v1.0.2

7: UniProt Protein Sequence Retrieval in R - https://github.com/virtual-labs/exp-sequence-retrieval-au, Master Branch, v1.0.1

8:  Gene finding: Finding Start and Stop codons using R - https://github.com/virtual-labs/exp-gene-finding-au, Master Branch, v1.0.1


<br>

**List of proposed experiments to be updated under Bioinformatics and Data Science in Biotechnology:** _(require server support)_
1: Retrieving a list of sequences from UniProt - https://github.com/virtual-labs/exp-uniprot-au, Main Branch, v1.0.0
2: Computing scoring matrices for amino acids and long pairwise alignment in R - https://github.com/virtual-labs/exp-long-pairwise-alignment-au Main Branch, v1.0.0
3: Calculating genetic distances between protein sequences - https://github.com/virtual-labs/exp-protein-sequences-au, Main Branch, v1.0.0
4: Differential expression analyses of RNA-seq - https://github.com/virtual-labs/exp-rna-seq-au, Main Branch, v1.0.0

**Comment by pavanchow** [comment]
@BalammaBoya 

Hosted Bioinformatics and Data Science in Biotechnology Lab with above experiments
URL of Lab: https://bds-au.vlabs.ac.in/
Branch: Master

Please check

**Comment by amritamindbrain** [comment]
@BalammaBoya @pavanchow we have updated the remaining experiments. Proceed with onboarding.

**Comment by pavanchow** [comment]
@amritamindbrain Could you please comment experiment name, URL, and tag of the experiments that need to be hosted?

**Comment by amritamindbrain** [comment]
1: Retrieving a list of sequences from UniProt - https://github.com/virtual-labs/exp-uniprot-au, Main Branch, v1.0.1

2: Computing scoring matrices for amino acids and long pairwise alignment in R - https://github.com/virtual-labs/exp-long-pairwise-alignment-au Main Branch, v1.0.1

3: Calculating genetic distances between protein sequences - https://github.com/virtual-labs/exp-protein-sequences-au, Main Branch, v1.0.1

4: Differential expression analyses of RNA-seq - https://github.com/virtual-labs/exp-rna-seq-au, Main Branch, v1.0.1

**Comment by pavanchow** [comment]
@amritamindbrain 

Hosted Bioinformatics and Data Science in Biotechnology Lab with mentioned experiments
URL of Lab: https://bds-au.vlabs.ac.in/
Branch: Master

Please check and approve

**Comment by amritamindbrain** [comment]
Checked. Please go ahead

**Comment by ravikiran2020** [comment]
Rehosting the lab with Bug Report and Rating Feature

**Comment by ravikiran2020** [comment]
Re-hosting the lab
